### DISCUSSION TITLE
AssemblyScript module with Emscripten? Possible? + share memory?

### DISCUSSION THREAD

**Discussion by alienself** [emscripten-core/emscripten#18670] [General]
Hi!

I want to load a wasm module compiled by AssemblyScript into a compiled wasm emscripten module written in c++, the goal would be to dynamically recompile the AssemblyScript module and allow the user to reload the module without recompiling the c++ module. Additionally both modules should share the same memory. 

How can I do that? Is this even possible?

**Comment by kripken** [comment]
If you can get the AssemblyScript module to use an imported memory, this could work, but you'd also need to get it to use a specified range of linear memory that the main program mallocs first (so they don't trample each other), and likewise for the table. That basically means doing something like dynamic linking manually. It's possible but it would be good to check with AssemblyScript if they have plans to support proper dynamic linking, in which case things would just work.

Alternatively, if AssemblyScript used Wasm GC then it would not need linear memory at all, and this would get a lot simpler.

cc @dcodeIO

**Reply by alienself** [reply]
Thank you for your reply, do you think that this could eventually be supported? 

My use case would be to write the core of my app in C++ and have a scripting capability that allows user to write in AssemblyScript and access exposed C++ classes from the core app module. 
I don't want to use JS directly for the scripting part because the memory would not be shared and the interop overhead too high. 

Perhaps AssemblyScript is not the right fit for this? Are you aware of any other project that would play well with wasm/emscripten?

**Reply by kripken** [reply]
If you want something that works right now, then I guess any language that can compile to a dynamic library could work. In theory that should be anything that compiles to LLVM as it can emit such libraries, but the only ones I'm aware of that have serious testing are C and C++ (in the emscripten test suite), which aren't really "scripting" languages... but there are some script-like languages that compile to C and C++, like [Lobster](https://strlen.com/lobster/).
